repo/content/list_cache.go

```diff
@@ -18,70 +18,69 @@ import (
 
 type listCache struct {
 	st                blob.Storage
-	cacheFile         string
+	cacheFilePrefix   string
 	listCacheDuration time.Duration
 	hmacSecret        []byte
 }
 
-func (c *listCache) listIndexBlobs(ctx context.Context) ([]IndexBlobInfo, error) {
-	if c.cacheFile != "" {
-		ci, err := c.readContentsFromCache(ctx)
+func (c *listCache) listIndexBlobs(ctx context.Context, prefix blob.ID) ([]blob.Metadata, error) {
+	if c.cacheFilePrefix != "" {
+		ci, err := c.readBlobsFromCache(ctx, prefix)
 		if err == nil {
 			expirationTime := ci.Timestamp.Add(c.listCacheDuration)
 			if time.Now().Before(expirationTime) { // allow:no-inject-time
-				log(ctx).Debugf("retrieved list of index blobs from cache")
-				return ci.Contents, nil
+				log(ctx).Debugf("retrieved list of %v '%v' index blobs from cache", len(ci.Blobs), prefix)
+				return ci.Blobs, nil
 			}
 		} else if err != blob.ErrBlobNotFound {
 			log(ctx).Warningf("unable to open cache file: %v", err)
 		}
 	}
 
-	contents, err := listIndexBlobsFromStorage(ctx, c.st)
+	blobs, err := blob.ListAllBlobs(ctx, c.st, prefix)
```

Comment:
We are now not calling ListAllBlobsConsistent. Correct me if these statements are not true -
1. This has not affect on listing for GCS, S3 because their list APIs gives a point in time list of blob IDs. (I agree that deleted blobs might also be present in the list. But it can never happen that two blobs A and B were written one after the other, and the list returns blob B but not blob A)
2. In case of local filesystem, the case I mentioned above where blob ID of B is listed but not that of A, can occur.

Reply:
I think if storage is strongly consistent (filesystem, GCS, but not S3), it's guaranteed that if you do PUT(a), PUT(b), LIST() will always return a & b.

S3 LIST on the other hand can return (in rare cases, but still):

[]
[a]
[b]
[a,b]

and for some (short) time after initial writes all of these are valid responses and they can fluctuate.

ListAllBlobsConsistent was listing until it received the same set of results, which is not really correct with S3 (because it can settle on [a] or [b] which are not really complete.